models/scheduled_pull_request_merge.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021 Gitea. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"code.gitea.io/gitea/modules/timeutil"
+)
+
+// ScheduledPullRequestMerge represents a pull request scheduled for merging when checks succeed
+type ScheduledPullRequestMerge struct {
+	ID          int64              `xorm:"pk autoincr"`
+	PullID      int64              `xorm:"BIGINT"`
+	DoerID      int64              `xorm:"BIGINT"`
+	Doer        *User              `xorm:"-"`
+	MergeStyle  MergeStyle         `xorm:"varchar(50)"`
+	Message     string             `xorm:"TEXT"`
+	CreatedUnix timeutil.TimeStamp `xorm:"created"`
+}
+
+// ScheduleAutoMerge schedules a pull request to be merged when all checks succeed
+func ScheduleAutoMerge(doer *User, pullID int64, style MergeStyle, message string) (err error) {
+	sess := x.NewSession()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+	defer sess.Close()
+
+	// Check if we already have a merge scheduled for that pull request
+	exists, _, err := getScheduledPullRequestMergeByPullID(sess, pullID)
+	if err != nil {
+		return
+	} else if exists {
+		return ErrPullRequestAlreadyScheduledToAutoMerge{PullID: pullID}
+	}
+
+	if _, err = sess.Insert(&ScheduledPullRequestMerge{
+		DoerID:     doer.ID,
+		PullID:     pullID,
+		MergeStyle: style,
+		Message:    message,
+	}); err != nil {
+		return
+	}
+
+	pr, err := getPullRequestByID(sess, pullID)
+	if err != nil {
+		return err
+	}
+
+	if _, err := createAutoMergeComment(sess, CommentTypePRScheduledToAutoMerge, pr, doer); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+// GetScheduledPullRequestMergeByPullID gets a scheduled pull request merge by pull request id
+func GetScheduledPullRequestMergeByPullID(pullID int64) (bool, *ScheduledPullRequestMerge, error) {
+	return getScheduledPullRequestMergeByPullID(x, pullID)
+}
+
+func getScheduledPullRequestMergeByPullID(e Engine, pullID int64) (bool, *ScheduledPullRequestMerge, error) {
+	scheduledPRM := &ScheduledPullRequestMerge{}
+	exists, err := e.Where("pull_id = ?", pullID).Get(scheduledPRM)
+	if err != nil || !exists {
+		return false, nil, err
+	}
+
+	doer, err := getUserByID(e, scheduledPRM.DoerID)
+	if err != nil {
+		return false, nil, err
+	}
+
+	scheduledPRM.Doer = doer
```

Comment:
not suggest do it here, suggest add ``func (p *ScheduledPullRequestMerge)LoadDoer() error``, and call it when needed.

Reply:
I dont see value to make an extra function if you mostly will load it anyway ... but **externaly**

if we would make a public LoadDoer it would be queryed in a seperate session, I like to avoid that

Reply:
only when need use doer data (like show it on ui) should we load doer data, or it's useless.
example change:

```DIFF
diff --git a/models/scheduled_pull_request_merge.go b/models/scheduled_pull_request_merge.go
index 1951a061e..f76e044e3 100644
--- a/models/scheduled_pull_request_merge.go
+++ b/models/scheduled_pull_request_merge.go
@@ -28,6 +28,7 @@ func ScheduleAutoMerge(doer *User, pullID int64, style MergeStyle, message strin
 	defer sess.Close()
 
 	// Check if we already have a merge scheduled for that pull request
+	// not need load doer here
 	if exists, _, err := getScheduledPullRequestMergeByPullID(sess, pullID); err != nil {
 		return err
 	} else if exists {
@@ -67,13 +68,26 @@ func getScheduledPullRequestMergeByPullID(e Engine, pullID int64) (bool, *Schedu
 		return false, nil, err
 	}
 
-	doer, err := getUserByID(e, scheduledPRM.DoerID)
+	return true, scheduledPRM, nil
+}
+
+func (scheduledPRM *ScheduledPullRequestMerge) LoadDoer() error {
+	if scheduledPRM.Doer != nil {
+		return nil
+	}
+
+	var err error 
+	scheduledPRM.Doer, err = GetUserByID(scheduledPRM.DoerID)
 	if err != nil {
-		return false, nil, err
+		if IsErrUserNotExist(err) {
+			scheduledPRM.Doer = NewGhostUser()
+			return nil
+		}
+
+		return err
 	}
 
-	scheduledPRM.Doer = doer
-	return true, scheduledPRM, nil
+	return nil
 }
 
 // RemoveScheduledPullRequestMerge cancels a previously scheduled pull request
@@ -84,6 +98,7 @@ func RemoveScheduledPullRequestMerge(doer *User, pullID int64, comment bool) err
 	}
 	defer sess.Close()
 
+	// not need load doer here
 	exist, scheduledPRM, err := getScheduledPullRequestMergeByPullID(sess, pullID)
 	if err != nil {
 		return err
diff --git a/routers/repo/issue.go b/routers/repo/issue.go
index be5651799..207145ad6 100644
--- a/routers/repo/issue.go
+++ b/routers/repo/issue.go
@@ -1568,11 +1568,21 @@ func ViewIssue(ctx *context.Context) {
 		ctx.Data["StillCanManualMerge"] = stillCanManualMerge()
 
 		// Check if there is a pending pr merge
-		ctx.Data["HasPendingPullRequestMerge"], ctx.Data["PendingPullRequestMerge"], err = models.GetScheduledPullRequestMergeByPullID(pull.ID)
+		hasPendingPullRequestMerge, pendingPullRequestMerge, err := models.GetScheduledPullRequestMergeByPullID(pull.ID)
 		if err != nil {
 			ctx.ServerError("GetScheduledPullRequestMergeByPullID", err)
 			return
 		}
+
+		if hasPendingPullRequestMerge {
+			err = pendingPullRequestMerge.LoadDoer()
+			if err != nil {
+				ctx.ServerError("pendingPullRequestMerge.LoadDoer", err)
+				return
+			}
+		}
+		ctx.Data["HasPendingPullRequestMerge"] = hasPendingPullRequestMerge
+		ctx.Data["PendingPullRequestMerge"] = pendingPullRequestMerge
 	}
 
 	// Get Dependencies
diff --git a/services/automerge/pull_auto_merge.go b/services/automerge/pull_auto_merge.go
index 0ad15d8bd..0b4277db8 100644
--- a/services/automerge/pull_auto_merge.go
+++ b/services/automerge/pull_auto_merge.go
@@ -125,12 +125,14 @@ func handlePull(pr *models.PullRequest, sha string) {
 	}
 
 	// Merge if all checks succeeded
-	doer, err := models.GetUserByID(scheduledPRM.DoerID)
+	err = scheduledPRM.LoadDoer()
 	if err != nil {
 		log.Error(err.Error())
 		return
 	}
 
+	// FIXME: should we still allow merge if doer is not exist
+
 	if err = pr.LoadBaseRepo(); err != nil {
 		log.Error(err.Error())
 		return
@@ -143,7 +145,7 @@ func handlePull(pr *models.PullRequest, sha string) {
 	}
 	defer baseGitRepo.Close()
 
-	if err := pullservice.Merge(pr, doer, baseGitRepo, scheduledPRM.MergeStyle, scheduledPRM.Message); err != nil {
+	if err := pullservice.Merge(pr, scheduledPRM.Doer, baseGitRepo, scheduledPRM.MergeStyle, scheduledPRM.Message); err != nil {
 		log.Error(err.Error())
 		return
 	}

```